### ISSUE TITLE
Elpy in Tramp buffers

### ISSUE THREAD

**Issue by jorgenschaefer** [jorgenschaefer/elpy#276]
Elpy currently does not work well in tramp (remote) buffers. As @UnwashedMeme summarized in #170:

> Elpy currently only runs python processes locally(1). When you are working on a remote tramp buffer it tries to start the elpy-rpc backend on the localhost with a `default-directory` that might be remote (not exactly sure how that's working) with the following observed behavior for `elpy-rpc-backend`:
> - `"jedi"` seems to do alright, but offers completions based on locally installed packages, not what's on the remote
> - `"rope"` chokes with the error of the ticket title
> - `"native"` works fine but has the limitations of native.
> - `nil` might detect rope on the localhost and err.
> ---
> 
> (1) At a minimum, to make this more tramp aware, https://github.com/jorgenschaefer/elpy/blob/master/elpy.el#L2274 needs to be changed to [start-file-process](https://www.gnu.org/software/emacs/manual/html_node/tramp/Remote-processes.html) and have the libraries installed remotely.

I am not sure what the correct behavior for Elpy would be. It can be one of two:

1) Elpy treats the remote buffer like a local buffer, using a local Python installation and local Python modules. This is tricky, because Elpy would need to find a remote project root, and backends likely dislike that (that's what Rope is annoyed at, I think). I'm not sure at all what the correct behavior here would be.

2) Elpy uses `start-file-proecess` as suggested. Then the helper functions will have to deal with remote buffers, telling the user that remote buffers require an Elpy installation on the remote system. A lot of complexity for a niche use case.

I'm not sure what the correct behavior should be, so I'll leave it at the current (suboptimal) behavior for now. If you have suggestions on how this should be handled, please do not hesitate to comment.

**Comment by bwo** [comment]
I don't think this is niche---I'm developing right now in a vagrant box where I have lots of non-python dev dependencies, and elpy really does not like this. I've kind of tricked it in some respects by setting the python rpc command to be a local shell script that sshs into the box and loads my venv and executes python there, but for lots of features (test running, etc.) it's a PITA as it stands.

**Comment by jorgenschaefer** [comment]
I am sorry this is a problem for you. Tramp uses a very serious amount of hacks and kludges to make things work like this, and it is not trivial for Elpy to add another layer of hacks on top of tramp to make this work. As I do not use this kind of setup, it is difficult for me to fix. If you feel this is an important feature, please do implement it and provide a pull request! Thank you :-)
